docs/proposals/volume-provisioning.md

```diff
@@ -0,0 +1,315 @@
+following behavior for this black box:
+
+1.  If multiple storage classes match the claim's selector, choose one at
+    random
+2.  If no storage class matches the claim's selector, use a default
+    class, which is specified as an argument to the controller manager
+
+### Controller workflow for provisioning volumes
+
+The workflow for provisioning does not need to change much to use storage
+classes.  The new workflow will be:
+
+1.  When a new claim is submitted, the controller attempts to find an existing
+    volume that will fulfill the claim.  If a suitable volume is found, the
+    controller binds the claim and volume together and carries on normally.
+2.  If no volume is found for the claim, the controller will attempt to
+    determine a storage class for the volume
+3.  If no storage class is found, the controller eventually retries finding
+    a storage class
+4.  All provisioners are in-tree; they implement an interface called
+    `ProvisionableVolumePlugin`, which has a method called `NewProvisioner`
+    that returns a new provisioner.
+5.  The provisioner's `Provision` method has the same responsibility as it
+    does now, but it is now passed both the claim and storage class as
+    parameters
+6.  If `Provision` returns an error, the controller stops trying to
+    provision
+7.  If `Provision` returns no error, the controller creates the returned
+    `api.PersistentVolume`, already bound to the claim
+  1.  If the create operation for the `api.PersistentVolume` fails, it is
+      retried
+  2.  If the create operation never succeeds, the controller attempts to
+      delete the provisioned volume and creates an event on the claim
+
+## Proposed Design
+
+We propose that:
+
+1.  A new API group called `storage` be created to hold the a `StorageClass`
+    API resource
+2.  The persistent volume controller be modified to dynamically provision
+    volumes using storage classes
+3.  The existing provisioner plugin implementations be modified to allow
+    parameterization as appropriate via storage classes
+4.  The existing alpha dynamic provisioning feature be phased out in the
+    next release
+
+### `StorageClass` API
+
+A new API group should hold the API for storage classes, following the pattern
+of autoscaling, metrics, etc.  To allow for future storage-related APIs, we
+should call this new API group `storage`.
+
+Storage classes will be represented by an API object called `StorageClass`:
+
+```go
+package storage
+
+// StorageClass describes the parameters for a class of storage for
+// which PersistentVolumes can be dynamically provisioned.
+//
+// StorageClasses are non-namespaced; the name of the storage class
+// according to etcd is in ObjectMeta.Name.
+type StorageClass struct {
+  unversioned.TypeMeta `json:",inline"`
+  ObjectMeta           `json:"metadata,omitempty"`
+
+  // ProvisionerType indicates the type of the provisioner.
+  ProvisionerType ProvisionerType `json:"provisionerType,omitempty"`
```

Comment:
I wouldn't enumerate these.  I would just make it an optional string which, if provided, indicates which driver to try to find and try to call.  Both of those steps might fail.  If not provided, this is implicitly an external provisioner.